ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,217 @@
+use std::collections::HashMap;
+
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, boolean_array::BooleanArray, ArrayAccess},
+    helpers::TotalRecords,
+    protocol::{
+        boolean::step::{EightBitStep, ThirtyTwoBitStep},
+        context::{
+            dzkp_validator::{DZKPValidator, TARGET_PROOF_SIZE},
+            Context, DZKPUpgraded, MaliciousProtocolSteps, UpgradableContext,
+        },
+        hybrid::step::HybridStep,
+        ipa_prf::boolean_ops::addition_sequential::integer_add,
+        BooleanProtocols, RecordId,
+    },
+    report::hybrid::{AggregateableHybridReport, PrfHybridReport},
+    secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, SharedValue},
+};
+
+pub async fn aggregate_reports<BK, V, C>(
+    ctx: C,
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Result<Vec<AggregateableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext,
+    BK: SharedValue + BooleanArray,
+    V: SharedValue + BooleanArray,
+    Replicated<Boolean>: BooleanProtocols<DZKPUpgraded<C>>,
+{
+    let mut reports_by_matchkey = HashMap::new();
+    reports_by_matchkey.reserve(reports.len() / 2);
+
+    for report in reports {
+        reports_by_matchkey
+            .entry(report.match_key)
+            .or_insert_with(Vec::new)
+            .push(report);
+    }
+
+    // an honest client and report collector will only submit
+    // one report with a breakdown key and one report with a value.
+    // if there are less, it's unattributed. if more, something went wrong.
+    // we remove these (instead of erroring/panicing).
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
+        .into_iter()
+        .filter_map(|(_, value)| {
+            if value.len() == 2 {
+                Some([value[0].clone().into(), value[1].clone().into()])
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let mut agg_reports = Vec::new();
+
+    let chunk_size = TARGET_PROOF_SIZE;
+
+    let dzkp_validator = ctx.clone().dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        std::cmp::min(ctx.active_work().get(), chunk_size.next_power_of_two()),
+    );
+
+    let ctx = dzkp_validator
+        .context()
+        .set_total_records(TotalRecords::specified(2 * report_pairs.len())?);
+
+    for (i, reports) in report_pairs.into_iter().enumerate() {
+        let record_id_bk = RecordId::FIRST + 2 * i;
+        let record_id_v = RecordId::FIRST + 2 * i + 1;
+        let (breakdown_key, _) = integer_add::<_, EightBitStep, 1>(
+            ctx.clone(),
```

Comment:
you need to narrow here and that shouldn't require record id math: https://github.com/private-attribution/ipa/blob/29578dd8bee2a274b194e3e2620b69aabf49f4e7/ipa-core/src/protocol/ipa_prf/aggregation/mod.rs#L188

Reply:
ty, let me give this a try. do you think that's the issue will address the stall?

Reply:
I think stall occurs because of not using seq_join